drivers/platform/stm32/stm32_xspi.c

```diff
@@ -0,0 +1,722 @@
+	if (!sdesc) {
+		ret = -ENOMEM;
+		goto sdesc_alloc_fail;
+	}
+
+	/* Allocate memory for STM32 specific platform XSPI command */
+	xspi_cmd = no_os_calloc(1, sizeof(*xspi_cmd));
+	if (!xspi_cmd) {
+		ret = -ENOMEM;
+		goto xspi_cmd_alloc_fail;
+	}
+
+	/* Copy settings to device descriptor */
+	xspi_desc->device_id = param->device_id;
+	xspi_desc->max_speed_hz = param->max_speed_hz;
+	xspi_desc->extra = sdesc;
+	xspi_desc->platform_ops = param->platform_ops;
+
+	/* Assign the command pointer with the allocated memory */
+	sdesc->command = xspi_cmd;
+
+	/* Get the input clock for the XSPI */
+	if (sinit->get_input_clock) {
+		sdesc->input_clock = sinit->get_input_clock();
+	}
+
+	sdesc->sample_shifting = sinit->sample_shifting;
+	sdesc->cs_high_time = sinit->cs_high_time;
+	sdesc->fifo_threshold = sinit->fifo_threshold;
+
+	/* Initialize the XSPI peripheral */
+	ret = stm32_xspi_peripheral_init(xspi_desc, param);
+	if (ret)
+		goto error;
+
+	/* Configure the XSPI with command */
+	ret = stm32_xspi_configure(xspi_desc, param);
+	if (ret)
+		goto error;
+
+	/* If DMA for XSPI exists in init params, initialize DMA */
+	if (sinit->dma_init) {
+		ret = no_os_dma_init(&sdesc->dma_desc, sinit->dma_init);
+		if (ret)
+			goto error;
+
+		if (sinit->dma_ch) {
+			sdesc->dma_ch = &sdesc->dma_desc->channels[0];
+			sdesc->dma_ch->id = sinit->dma_ch->hdma;
+			sdesc->dma_ch->extra = sinit->dma_ch;
+			sdesc->dma_ch->irq_num = sinit->irq_num;
+		}
+	}
+
+	*desc = xspi_desc;
+
+	return 0;
+
+error:
+	no_os_dma_remove(sdesc->dma_desc);
+	no_os_free(xspi_cmd);
+xspi_cmd_alloc_fail:
+	no_os_free(sdesc);
+sdesc_alloc_fail:
+	no_os_free(xspi_desc);
+xspi_desc_alloc_fail:
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by no_os_spi_init().
+ * @param desc - The XSPI descriptor.
+ * @return 0 in case of success, negative code otherwise.
+ */
+static int32_t stm32_xspi_remove(struct no_os_spi_desc *desc)
+{
+	struct stm32_xspi_desc *sdesc;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	sdesc = desc->extra;
+
+	/* Remove the DMA descriptor */
+	no_os_dma_remove(sdesc->dma_desc);
+
+	/* De-Initialize XSPI peripheral */
+	HAL_XSPI_DeInit(&sdesc->hxspi);
+
+	/* Free the command */
+	no_os_free(sdesc->command);
+
+	/* Free the platform specific descriptor */
+	no_os_free(sdesc);
+
+	/* Free the no-os descriptor */
+	no_os_free(desc);
+
+	return 0;
+}
+
+/**
+ * @brief Write/read multiple messages to/from XSPI.
+ * @param desc - The XSPI descriptor.
+ * @param msgs - The messages array.
+ * @param len - Number of messages.
+ * @return 0 in case of success, negative code otherwise.
+ */
+static int32_t stm32_xspi_transfer(struct no_os_spi_desc *desc,
+				   struct no_os_spi_msg *msgs,
+				   uint32_t len)
+{
+	int32_t ret;
+	struct stm32_xspi_desc *sdesc;
+	XSPI_RegularCmdTypeDef *cmd;
+
+	if (!desc || !desc->extra || !msgs)
+		return -EINVAL;
+
+	sdesc = desc->extra;
+	cmd = sdesc->command;
+
+	for (uint32_t i = 0; i < len; i++) {
```

Comment:
For the standard SPI peripheral we're configuring the hardware registers before the transfer (based on the fields in the no os and STM32 specific SPI descriptors). I think we could do the same in this case and avoid the need to have the new `configure` API.

Reply:
The original intent for adding configure API is to reconfigure the vendor specific params without reinitializing the peripheral and creating descriptors again. If this is the case, instead of `no_os_spi_init_params` as argument, is it good to have `void *params` just to reconfigure vendor specific params?

If configuration need to be done, then user will use `no_os_spi_init` with new set of init params.

Reply:
I think you could do what @CiprianRegus says for now and drop it. A configure function would be very useful but not necessarily how it's implemented. The way I see it is that for each peripheral driver we could have:
- init() - allocate memory, calls config() internally
- config() - configure the peripheral
- remove() - deallocate memory (and put the hardware in a good state, if needed)

The user could then reconfigure the peripheral with a new set of init_param structure without deallocating/reallocating memory. 

But this is quite a big change and I don't think this PR should do such big changes, just drop the config function for now and we'll see about it in the future.

Reply:
PS: the idea was good, we even have an issue with the UART where we'd like to reconfigure the baudrate without reallocating it, like you proposed for this spi, it's just that we need to have a proper discussion about this internally